Refactored JobResultUIViewSet

Pavan-183 opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
# Closes #<ISSUE NUMBER GOES HERE>
# What's Changed
<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
-->
1. As JobResult doesn't support edit functionality, I haven't used `NautobotUIViewSet`, instead I inherited only required mixins.
2. Created a new file named `jobresult_retrieve.html` by copying the contents of the original `jobresult.html`.

# Screenshots
<!--
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->
| View             | Before                                                                                                                         | After|
|------------------|--------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------|
| List View        | ![image](https://github.com/user-attachments/assets/b98bef73-5d5b-4b86-9963-8a3134520ce2) <br> ![image](https://github.com/user-attachments/assets/ac11dadf-40a3-4caf-8cf0-1ec9cf5f581d) | ![image](https://github.com/user-attachments/assets/46a30c72-0fd5-405c-a890-897a32f9b302) <br> ![image](https://github.com/user-attachments/assets/5dbd836a-06dd-4cd6-8657-58e17930e8e4) |
| Detail View      | ![image](https://github.com/user-attachments/assets/c851d801-ed1d-4b59-b29f-17029e0dcd3e) | ![image](https://github.com/user-attachments/assets/0a0e5f84-d7ed-4520-a21f-27a21cc7a1ea) |
| Update View      | Not available | Not available |
| Delete View      | ![image](https://github.com/user-attachments/assets/8829c5a3-2ddf-404b-be22-927dc1cb8d64) <br> ![image](https://github.com/user-attachments/assets/f40d4a3d-687d-4889-8936-4986fcf26591) | ![image](https://github.com/user-attachments/assets/76f8d094-c41d-4721-8e92-21cc6d01fb05) <br> ![image](https://github.com/user-attachments/assets/4bef2f8c-7026-4b52-8fb9-a682cebbcc3c) |
| Bulk Edit View   | Not available | Not available |
| Bulk Delete View | ![image](https://github.com/user-attachments/assets/b9a14d5b-c010-4562-af1b-7605e20f0548) <br> ![image](https://github.com/user-attachments/assets/f30237fb-be2d-4367-9410-f514e9ef6401) | ![image](https://github.com/user-attachments/assets/f1b78b5f-62ff-4064-901d-06d54b5d7f49) <br> ![image](https://github.com/user-attachments/assets/6556b5df-e275-4d7f-a272-e7d17c6cf628) |



# TODO
<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

Sridevi-NTC commented on nautobot/extras/urls.py:
```diff
@@ -27,6 +27,7 @@
 router.register("job-buttons", views.JobButtonUIViewSet)
 router.register("job-hooks", views.JobHookUIViewSet)
 router.register("job-queues", views.JobQueueUIViewSet)
+router.register("job-results", views.JobResultUIViewSet)
```
Seems like bulk edit and bulk delete are not available. Could you please add only required mixins?

Pavan-183 commented:
I will look into this @itdependsnetworks

glennmatthews commented on nautobot/extras/views.py:
```diff
@@ -2148,37 +2148,22 @@ class JobHookUIViewSet(NautobotUIViewSet):
+    filterset_form_class = forms.JobResultFilterForm
+    serializer_class = serializers.JobResultSerializer
+    table_class = tables.JobResultTable
     queryset = JobResult.objects.defer("result").select_related("job_model", "user")
```
`defer` is intentionally here as a performance improvement on the list views specifically. It should *not* be done on the detail views. Come to think of it, for the non-detail cases, it should probably be augmented to `.defer(result, task_args, task_kwargs, celery_kwargs, traceback, meta)` since all of those fields are potentially large and none are displayed in the list views.

The `select_related` is *probably* not needed as an explicit specification at this time.

glennmatthews replied:
I was thinking `if not self.detail` rather than `self.action != retrieve` but otherwise pretty much.

itdependsnetworks replied:
okay, did not know about self.detail, so good to know.

Pavan-183 commented:
```
    def get_table_class(self):
        if "delete" in self.request.path or "bulk_destroy" in self.request.path:
            # Create a subclass of the table with the "actions" column removed
            class JobResultTableWithoutActions(self.table_class):
                class Meta(self.table_class.Meta):
                    # Exclude the "actions" column from the field list
                    all_default_columns = self.table_class.Meta.default_columns
                    default_columns = tuple(column for column in all_default_columns if column != "actions")

            return JobResultTableWithoutActions

        return self.table_class
```

@itdependsnetworks , I’ve added the following logic to the `get_table_class()` method to remove the actions column during bulk delete confirmation.
Instead of changing the original `JobResultTable`, I fixed the issue by:

1. Dynamically creating a new version of the table inside the view
2. Skipping the "actions" column in that version
3. Using this version only when the user is doing a bulk delete

This way, the action buttons are hidden during bulk delete, and everything else stays the same. Please let me know if there's a better way to handle this or if I'm missing anything here.


| Before                                                                                     | After                                                                                      |
|--------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------|
| <img src="https://github.com/user-attachments/assets/c4f55786-676b-4043-841f-1a90be083fb6" width="300" /> | <img src="https://github.com/user-attachments/assets/10338a8d-8883-4b64-8862-eaa038b13a40" width="600" /> |

glennmatthews commented:
Bulk-delete is *supposed* to auto-hide the `actions` column normally - see around line 140 of `nautobot/core/views/renderers.py`. It shouldn't be necessary to create a custom table subclass.

Pavan-183 commented:
@glennmatthews 

I removed the `get_table_class()` override as you mentioned. But the "actions" column is still showing up in the bulk delete screen.
It looks like the column isn’t getting hidden as expected, maybe view.action isn’t set to "bulk_destroy" in this case. Still checking on that.
Please let me know if there’s anything else I should try!

Pavan-183 commented:
@glennmatthews Updated as requested, should be resolved, let me know if I missed anything.

glennmatthews commented on nautobot/extras/urls.py:
```diff
@@ -236,19 +237,7 @@
     # Generic job results
-    path("job-results/", views.JobResultListView.as_view(), name="jobresult_list"),
-    path("job-results/<uuid:pk>/", views.JobResultView.as_view(), name="jobresult"),
     path("job-results/<uuid:pk>/log-table/", views.JobLogEntryTableView.as_view(), name="jobresult_log-table"),
```
Can we convert this to a custom action on the viewset as well?

Pavan-183 replied:
Okay @glennmatthews

Pavan-183 commented on nautobot/core/views/mixins.py:
```diff
@@ -1142,6 +1142,9 @@ def perform_bulk_destroy(self, request, **kwargs):  # pylint:disable=arguments-r
                 return redirect(self.get_return_url(request))
+            # Hide actions column in the table for bulk destroy view
+            if "actions" in table.columns:
+                table.columns.hide("actions")
```
Updated as requested, should be resolved, let me know if I missed anything
